src/mon/OSDMonitor.h

```diff
@@ -124,6 +124,128 @@ class LastEpochClean {
 };
 
 
+struct osdmap_manifest_t {
+  // all the maps we have pinned -- i.e., won't be removed unless
+  // they are inside a trim interval.
+  set<version_t> pinned;
+  // the last pruned version
+  version_t last_pruned;
+
+  osdmap_manifest_t() : last_pruned(0) { }
+
+  version_t get_last_pinned() const
+  {
+    set<version_t>::const_reverse_iterator it = pinned.crbegin();
+    if (it == pinned.crend()) {
+      ceph_assert(pinned.empty());
+      return 0;
+    }
+    return *it;
+  }
+
+  version_t get_first_pinned() const
+  {
+    set<version_t>::const_iterator it = pinned.cbegin();
+    if (it == pinned.cend()) {
+      ceph_assert(pinned.empty());
+      return 0;
+    }
+    return *it;
+  }
+
+  version_t get_last_pruned() const
+  {
+    return last_pruned;
+  }
+
+  void set_last_pruned(version_t v) {
+    last_pruned = v;
+  }
+
+  bool is_pinned(version_t v) const
+  {
+    return pinned.find(v) != pinned.end();
+  }
+
+  void pin(version_t v)
+  {
+    pinned.insert(v);
+  }
+
+  version_t get_lower_closest_pinned(version_t v) const
+  {
+    set<version_t>::const_iterator p = pinned.lower_bound(v);
+    if (p == pinned.cend()) {
+      return 0;
+    }
+
+    if (*p > v) {
+      if (p == pinned.cbegin()) {
+        return 0;
+      }
+      --p;
+    }
+
+    return *p;
+  }
+
+  version_t get_upper_closest_pinned(version_t v) const
+  {
+    set<version_t>::const_iterator p = pinned.upper_bound(v);
+    if (p == pinned.cend()) {
+      return 0;
+    }
+    return *p;
+  }
+
+  /* obtain the next prunable version
+   */
+  version_t get_next_prunable() const
+  {
+    return get_next_prunable(last_pruned);
+  }
+
+  version_t get_next_prunable(version_t v) const
```

Comment:
I wonder if it would be simpler to just drop this last_pruned, in_progress thing.  Instead, when we prune, we (in a single transaction) (1) add the pinned epoch and (2) delete all maps between the new pinned epoch and the next one.  I don't think the gap should be more then a few 10s of epochs anyway, and deletes aren't expensive to include in a transaction.